### PR TITLE
UIRegistry now extensible, like other Registries

### LINKED ISSUE #2513: Partial use of common registry API forced by UIComponentRegistryFactory
Most of openHAB APIs, whether these are things, items or links rely on dedicated `Registry` instance feed by one or more `Provider` services. This design helps to have various sources of core elements managed in various ways.
This is however not reflected within UI registries for pages and widgets since they are wrapped and maintained only via storage.

I believe this is not a major trouble, however current construction does not permit ie. read only UI elements to be present in system. They can't be dynamically computed based on supplied configurations, present items and such.
Changing that would open several fields - ie. where pages could be provided by further system extensions.

### PR DESCRIPTION
**Pull request by jpg0**
This is to allow the UI registry & providers to follow the standard extensible pattern to allow osgi to supply other widgets and pages.

Fixes #2513 

Signed-off-by: Jonathan Gilbert <jpg@trillica.com>

### CODE REVIEW

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java
Comment (jpg0) [inline bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java:78]: @splatch I modified this section to make the collections immutable, as I was having issues with them not being (not sure why).

Comment (splatch) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java:78]: Making final set inserted into map immutable is wrong, cause you can't remove provider later on when it disappear. 

Given that component activation can be made through different threads (ie. declarative services, OSGi activator etc), you probably better of with synchronized collection rather than plain `HashSet`. Keep in mind that set registered for namespace will be manipulated.

**Comment by splatch** [comment]
@jpg0 Since you made modifications to initial code I posted feel free to add yourself to authors :)

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java
Comment (jpg0) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java:78]: Well, whilst the set was immutable, I replaced it with a completely new set when it was modified, so it did still work.

Saying this, I don't really like the verbose code to do this (Java mutable collections I guess). Instead I have now changed it to use a `CopyOnWriteArraySet`, and also synchronized write access (otherwise concurrent updates could lose elements), as it should happen very infrequently anyway.

**Review by wborn — DISMISSED** [review]
Thanks for making it extensible! I added a few minor comments below:

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIProvider.java
Comment (splatch) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIProvider.java:2]: Fair spot, but this is something which isn't really holding up this PR from going in could be addressed by automatic spotless fix once PR is merged (or while it is merged through robotized process). This is a generic point on multiple PRs which suffer from same issue as there is bunch of them for core and addons.
IANAL, but I need note that date in source file header is authoring date which is different for each file. Copyright can be established at time of making the sources, hence in theory it can't be "2010" for file which was made in 2021. It might be also doubtful if changing a comment which does not change contributed works. To make an example - re-printing of an book released in 1999 without changes in 2022 does not extend the copyright as the printing date is just an annotation on the covers, just like source file header.

Comment (wborn) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIProvider.java:2]: Sorry I normally don't merge PRs that will break builds.😉

**Comment by splatch** [comment]
> +/**
> + * Copyright (c) 2010-2021 Contributors to the openHAB project
> Sorry I normally don't merge PRs that will break builds.😉
> 
Maybe then its worth to consider changing header template to have `Copyright (c) 2010-\d+`. It will save tens of pull requests stuck for same plain reason. Especially these made in last two months of each year. Unless it is better to comment each of affected files individually?

**Comment by wborn** [comment]
I think it makes perfect sense to have the same headers everywhere.

> Unless it is better to comment each of affected files individually?

That's done automatically with GitHub annotations for new PRs nowadays.

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIProvider.java
Comment (wborn) [inline bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIProvider.java:22]: Typo:

```suggestion
 * @author Łukasz Dywicki - initial contribution
```

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryImpl.java
Comment (wborn) [inline bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryImpl.java:30]: Another typo: 

```suggestion
 * @author Łukasz Dywicki - Support for dynamic registration of providers.
```

**Comment by splatch** [comment]
> That's done automatically with GitHub annotations for new PRs nowadays.

Why not making automatic spotless:apply as a part of PR processing? If comments are automated, why fixes aren't?
It would save everybodys time and let people focus on issues which are not handled by tooling.

**Comment by wborn** [comment]
> If comments are automated, why fixes aren't?

It's a nice idea but adding commits to branches causes sync issues and conflicts which then need to be resolved. That's probably why most repos don't do this. I've seen more contributors struggle with Git than with fixing the trivial issues. :wink: Maybe a bot that can fix it on demand would work better like how @dependabot works.

File: bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java
Comment (J-N-K) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java:51]: Yes they are, because otherwise the null constraints are not fulfilled. The reason is that the whole class is not null-annotated. Since the other issues are only typos / year correction and we are allowed to edit the PR, are you ok with merging after the suggestions have been committed?

Comment (wborn) [reply bundles/org.openhab.core.ui/src/main/java/org/openhab/core/ui/internal/components/UIComponentRegistryFactoryImpl.java]: OK I've added the annotation on the class and done some more trivial cleanup/fixes.
LGTM now, so if you are happy with it as well, feel free to merge it. :slightly_smiling_face:

**Review by J-N-K — APPROVED** [review]

**Comment by J-N-K** [comment]
Thanks!

**Pull request merged**
